File: src/main/java/io/protobase/f7/formulas/statistical/COUNTA.java
package io.protobase.f7.formulas.statistical;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Mappers;

import java.util.Arrays;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

public class COUNTA extends AbstractFormula {
  public static FormulaName NAME = FormulaName.COUNTA;
  public static COUNTA SELF = new COUNTA();

  public COUNTA() {
    super();
  }

  public COUNTA(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkAtLeastLength(values.length, 1, NAME);
    return ((double) Arrays.stream(values)
        .map(lookup)
        .flatMap(Mappers::toFlatStream)
        .collect(Collectors.toList()).size());
  }
}

File: src/main/js/formulas/statistical/COUNTA.ts
import { isNotNull } from "../../utils/Other";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Counter } from "../../utils/Counter";
import { Mappers } from "../../utils/Mappers";
import { Predicates } from "../../utils/Predicates";
import { Reducers } from "../../utils/Reducers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class COUNTA extends AbstractFormula {
  static SELF: COUNTA = new COUNTA();
  NAME = FormulaName.COUNTA;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkAtLeastLength(values.length, 1, this.NAME);
    const counter = new Counter<Complex>();
    values
      .map(this.lookup)
      .map(Mappers.flattenGridsToArrays) // 1, 2, [null, 1, "0", "Text.", null, #REF!]
      .map(Mappers.ensureAllAreArrays) // [1], [2], [null, 1, "0", "Text.", null, #REF!]
      .reduce(Reducers.join) // [1, 2, null, 1, "0", "Text.", null, #REF!]
      .filter(isNotNull) // [1, 2, 1, "0", "Text.", #REF!]
      .filter(Predicates.isNonError) // [1, 2, 1, "0", "Text."]
      .map((value) => counter.count(value));
    return counter.getCount();
  }
}
